File: dock_spawn/tab/tab_handle.dart
part of dock_spawn;

/** 
 * A tab handle represents the tab button on the tab strip
 */
class TabHandle {
  static int zIndexCounter = 1000;
  DivElement elementBase;
  DivElement elementText;
  DivElement elementCloseButton;
  UndockInitiator undockInitiator;
  TabPage parent;
  var undockHandler;

  var mouseClickHandler;  // Button click handler for the tab handle
  var closeButtonHandler; // Button click handler for the close button
  
  TabHandle(this.parent) {
    undockHandler = _performUndock;
    elementBase = new DivElement();
    elementText = new DivElement();
    elementCloseButton = new DivElement();
    elementBase.classes.add("tab-handle");
    elementBase.classes.add("disable-selection"); // Disable text selection
    elementText.classes.add("tab-handle-text");
    elementCloseButton.classes.add("tab-handle-close-button");
    elementBase.nodes.add(elementText);
    if (parent.host.displayCloseButton) {
      elementBase.nodes.add(elementCloseButton);
    }
    
    parent.host.tabListElement.nodes.add(elementBase);
    
    PanelContainer panel = parent.container;
    String title = panel.getRawTitle();
    elementText.innerHtml = title;
    
    // Set the close button text (font awesome)
    String closeIcon = "icon-remove-sign";
    elementCloseButton.innerHtml = '<i class="$closeIcon"></i>';
    
    _bringToFront(elementBase);
    
    undockInitiator = new UndockInitiator(elementBase, undockHandler);
    undockInitiator.enabled = true;

    mouseClickHandler = onMouseClicked;
    elementBase.on.click.add(mouseClickHandler);
    
    closeButtonHandler = onCloseButtonClicked;
    elementCloseButton.on.mouseDown.add(closeButtonHandler);
    
  }
  
  void destroy() {
    elementBase.on.click.remove(mouseClickHandler);
    elementCloseButton.on.mouseDown.remove(closeButtonHandler);
    elementBase.remove();
    elementCloseButton.remove();
    elementBase = null;
    elementCloseButton = null;
  }

  Dialog _performUndock(MouseEvent e, Point2 dragOffset) {
    if (parent.container.containerType == "panel") {
      undockInitiator.enabled = false;
      PanelContainer panel = parent.container;
      return panel.performUndockToDialog(e, dragOffset);
    }
    else {
      return null;
    }
  }
  
  void onMouseClicked(MouseEvent e) {
    parent.onSelected();
  }
  
  void onCloseButtonClicked(MouseEvent e) {
    // If the page contains a panel element, undock it and destroy it
    if (parent.container.containerType == "panel") {
      undockInitiator.enabled = false;
      PanelContainer panel = parent.container;
      panel.performUndock();
    }
  }
  
  void setSelected(bool selected) {
    String selectedClassName = "tab-handle-selected";
    if (selected) {
      elementBase.classes.add(selectedClassName);
    } else {
      elementBase.classes.remove(selectedClassName);
    }
  }
  
  void setZIndex(int zIndex) {
    elementBase.style.zIndex = "$zIndex";
  }
  
  void _bringToFront(Element element) {
    element.style.zIndex = "$zIndexCounter";
    zIndexCounter++;
  }
}

File: js/lib/tab/TabHandle.js
/**
 * A tab handle represents the tab button on the tab strip
 */
dockspawn.TabHandle = function(parent)
{
    this.parent = parent;
    var undockHandler = dockspawn.TabHandle.prototype._performUndock.bind(this);
    this.elementBase = document.createElement('div');
    this.elementText = document.createElement('div');
    this.elementCloseButton = document.createElement('div');
    this.elementBase.classList.add("tab-handle");
    this.elementBase.classList.add("disable-selection"); // Disable text selection
    this.elementText.classList.add("tab-handle-text");
    this.elementCloseButton.classList.add("tab-handle-close-button");
    this.elementBase.appendChild(this.elementText);
    if (this.parent.host.displayCloseButton)
        this.elementBase.appendChild(this.elementCloseButton);

    this.parent.host.tabListElement.appendChild(this.elementBase);

    var panel = parent.container;
    var title = panel.getRawTitle();
    this.elementText.innerHTML = title;

    // Set the close button text (font awesome)
    var closeIcon = "icon-remove-sign";
    this.elementCloseButton.innerHTML = '<i class="' + closeIcon + '"></i>';

    this._bringToFront(this.elementBase);

    this.undockInitiator = new dockspawn.UndockInitiator(this.elementBase, undockHandler);
    this.undockInitiator.enabled = true;

    this.mouseClickHandler = new dockspawn.EventHandler(this.elementBase, 'click', this.onMouseClicked.bind(this));                     // Button click handler for the tab handle
    this.closeButtonHandler = new dockspawn.EventHandler(this.elementCloseButton, 'mousedown', this.onCloseButtonClicked.bind(this));   // Button click handler for the close button

    this.zIndexCounter = 1000;
};

dockspawn.TabHandle.prototype.updateTitle = function()
{
    if (this.parent.container instanceof dockspawn.PanelContainer)
    {
        var panel = this.parent.container;
        var title = panel.getRawTitle();
        this.elementText.innerHTML = title;
    }
};

dockspawn.TabHandle.prototype.destroy = function()
{
    this.mouseClickHandler.cancel();
    this.closeButtonHandler.cancel();
    removeNode(this.elementBase);
    removeNode(this.elementCloseButton);
    delete this.elementBase;
    delete this.elementCloseButton;
};

dockspawn.TabHandle.prototype._performUndock = function(e, dragOffset)
{
    if (this.parent.container.containerType == "panel")
    {
        this.undockInitiator.enabled = false;
        var panel = this.parent.container;
        return panel.performUndockToDialog(e, dragOffset);
    }
    else
        return null;
};

dockspawn.TabHandle.prototype.onMouseClicked = function()
{
    this.parent.onSelected();
};

dockspawn.TabHandle.prototype.onCloseButtonClicked = function()
{
    // If the page contains a panel element, undock it and destroy it
    if (this.parent.container.containerType == "panel")
    {
        this.undockInitiator.enabled = false;
        var panel = this.parent.container;
        panel.performUndock();
    }
};

dockspawn.TabHandle.prototype.setSelected = function(selected)
{
    var selectedClassName = "tab-handle-selected";
    if (selected)
        this.elementBase.classList.add(selectedClassName);
    else
        this.elementBase.classList.remove(selectedClassName);
};

dockspawn.TabHandle.prototype.setZIndex = function(zIndex)
{
    this.elementBase.style.zIndex = zIndex;
};

dockspawn.TabHandle.prototype._bringToFront = function(element)
{
    element.style.zIndex = this.zIndexCounter;
    this.zIndexCounter++;
};
